boot/freeldr/bootsect/pc98/fat12fdd.S

```diff
@@ -0,0 +1,579 @@
+ * Call with:
+ *
+ * AL - number
+ */
+Print8HEX:
+    push ax
+
+    shr al, 4                                   // Move upper nibble into lower nibble
+    call Print8HEXLowNibble
+
+    pop ax
+    push ax
+    call Print8HEXLowNibble
+
+    pop ax
+
+    ret
+
+Print8HEXLowNibble:
+    and al, HEX(0F)                             // Ignore upper nibble
+
+    cmp al, HEX(09)                             // 0Ah-0Fh or 00h-09h
+    jbe .ASCIISkip
+
+    add al, HEX(7)                              // Convert 0Ah-0Fh to ASCII 41h-46h
+
+    .ASCIISkip:
+        add al, HEX(30)                         // Convert 00h-09h to ASCII 30h-39h
+
+    call PutChar                                // Display it
+
+    ret
+
+/*
+ * Prints a HEX value of AX
+ *
+ * Call with:
+ *
+ * AX - number
+ */
+Print16HEX:
+    push ax
+
+    xchg ah, al                                 // High byte
+    call Print8HEX
+
+    xchg ah, al                                 // Low byte
+    call Print8HEX
+
+    pop ax
+
+    ret
+
+/*
+ * Displays a hardware error message and reboots
+ */
+HardwareError:
+    mov si, offset msgHardwareError
+
+    .PrintStringVGA:
+        lodsb                                   // Get a single char from a ptr
+
+        or al, al
+        jz short .HardwareErrorDone             // Found NULL
+
+        mov ah, HEX(0E)                         // Teletype output
+        mov bx, 7                               // BH - video page number, BL - foreground color
+        int HEX(10)                             // Display a character via TTY mode
+    jmp short .PrintStringVGA
+
+.HardwareErrorDone:
+    xor ax, ax
+    int HEX(16)                                 // Wait for a keypress
+    int HEX(19)                                 // Reboot
+endif
+
+/*
+ * Displays a disk error message and reboots
+ */
+PrintDiskError:
+if 0
+    mov al, byte ptr BP_REL(lastReadStatus)
+    call Print8HEX                              // Display the last read status
+endif
+
+    mov si, offset msgDiskError                 // Disk error message
+    call PrintString                            // Display it
+
+    jmp short Reboot
+
+/*
+ * Displays a file not found error message and reboots
+ */
+PrintFileNotFound:
+    mov si, offset msgFreeLdr                   // FreeLdr not found message
+    call PrintString                            // Display it
+
+    jmp short Reboot
+
+/*
+ * Reboots the computer after keypress
+ */
+Reboot:
+    mov si, offset msgAnyKey                    // Press any key message
+    call PrintString                            // Display it
+
+    xor ax, ax
+    int HEX(18)                                 // Wait for a keypress
+
+    /*
+     * Activate the CPU reset line
+     * See https://people.freebsd.org/~kato/pc98-arch.html#cpureset
+     * and http://www.webtech.co.jp/company/doc/undocumented_mem/io_cpu.txt
+     */
+    xor ax, ax
+    outb HEX(0F0), al
+
+    hlt
+    jmp short $ - 1                             // Spin
```

Comment:
I doubt this will compile.... (we are not using NASM!)

Reply:
Fixed.